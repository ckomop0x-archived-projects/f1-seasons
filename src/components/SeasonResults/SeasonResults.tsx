import {
  BackButton,
  FavoriteButton,
  StandingsStyled,
  StandingsTable,
  StandingsTableRow
} from './styles';
import { RacesResult, Result } from '../../types';
import { PureComponent } from 'react';

interface SeasonResultsProps {
  season: string;
  round: string;
  year: string;

  toggleViews(): void;
}

interface SeasonResultsState {
  isUpdating: boolean;
  racesResult?: RacesResult;
  savedDrivers: string[];
}

class SeasonResults extends PureComponent<SeasonResultsProps, SeasonResultsState> {
  constructor(props: SeasonResultsProps) {
    super(props);
    const savedDrivers: string = localStorage.getItem('savedDrivers') || '';

    this.state = {
      isUpdating: false,
      savedDrivers: savedDrivers ? JSON.parse(savedDrivers) : []
    };
  }

  // private load(): void {
  //   if (this.state.isUpdating) {
  //     return;
  //   }
  //   this.setState({ isUpdating: true });
  //   getSeasonsResults(this.props.season, this.props.round)
  //     .then((data) => {
  //       this.setState({
  //         racesResult: data.MRData.RaceTable.Races[0],
  //         isUpdating: false
  //       });
  //     })
  //     .catch((error) => {
  //       alert(error);
  //       this.setState({
  //         isUpdating: false
  //       });
  //     });
  // }

  addDriverToFavorite = (driverCode: string): void => {
    this.setState({
      savedDrivers: this.state.savedDrivers.concat(driverCode)
    });
  };

  removeDriverFromFavorites = (driverCode: string): void => {
    this.setState({
      savedDrivers: this.state.savedDrivers.filter(
        (savedDriverCode): boolean => savedDriverCode !== driverCode
      )
    });
  };

  saveFavoritesDrivers(): void {
    localStorage.setItem('savedDrivers', JSON.stringify(this.state.savedDrivers));
  }

  render() {
    return (
      <StandingsStyled>
        <BackButton onClick={this.props.toggleViews}>&larr; Back to seasons</BackButton>
        {this.state.isUpdating && (
          <div style={{ textAlign: 'center', marginTop: 40 }}>Loading results...</div>
        )}
        {this.state.racesResult && <h2>{this.state.racesResult.Circuit.circuitName}</h2>}
        {this.state.racesResult ? (
          <StandingsTable>
            <table>
              <thead>
                <tr>
                  <th>Position</th>
                  <th>Number</th>
                  <th>Driver</th>
                  <th>Code</th>
                  <th>Favorites</th>
                </tr>
              </thead>
              <tbody>
                {this.state.racesResult &&
                  this.state.racesResult.Results.map((result: Result) => {
                    const { Driver } = result;
                    const isDriverFavorite: boolean = this.state.savedDrivers.includes(Driver.code);

                    return (
                      <StandingsTableRow
                        key={Driver.code}
                        position={result.position}
                        onClick={
                          isDriverFavorite
                            ? this.removeDriverFromFavorites.bind(this, Driver.code)
                            : this.addDriverToFavorite.bind(this, Driver.code)
                        }>
                        <td>{result.position}</td>
                        <td>{result.number}</td>
                        <td>
                          {Driver.givenName} {Driver.familyName}
                        </td>
                        <td>{Driver.code}</td>
                        <td>
                          <FavoriteButton isDriverFavorite={isDriverFavorite} />
                        </td>
                      </StandingsTableRow>
                    );
                  })}
              </tbody>
            </table>
          </StandingsTable>
        ) : (
          !this.state.isUpdating && (
            <div style={{ textAlign: 'center', marginTop: 40 }}>No results available</div>
          )
        )}
      </StandingsStyled>
    );
  }
}

export default SeasonResults;
