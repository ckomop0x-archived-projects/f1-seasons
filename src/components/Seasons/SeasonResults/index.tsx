import { PureComponent } from 'react';

import getSeasonsResults from '../../../services/get-seasons-results';
import { RacesResult, Result } from '../types';

import {
  BackButton,
  FavoriteButton,
  StandingsStyled,
  StandingsTable,
  StandingsTableRow,
} from './styles';

interface SeasonResultsProps {
  season: string;
  round: string;
  year: string;
  toggleViews(): void;
}

interface SeasonResultsState {
  isUpdating: boolean;
  racesResult?: RacesResult;
  savedDrivers: string[];
}

class SeasonResults extends PureComponent<
  SeasonResultsProps,
  SeasonResultsState
> {
  private loadingTimerId: any;

  constructor(props: SeasonResultsProps) {
    let savedDrivers: string[] = [];
    super(props);
    if (typeof window !== 'undefined') {
      savedDrivers =
        JSON.parse(localStorage.getItem('savedDrivers') || '') || [];
    }

    this.state = {
      isUpdating: false,
      savedDrivers,
    };
  }

  addDriverToFavorite = (driverCode: string): void => {
    this.setState({
      savedDrivers: this.state.savedDrivers.concat(driverCode),
    });
  };

  removeDriverFromFavorites = (driverCode: string): void => {
    this.setState({
      savedDrivers: this.state.savedDrivers.filter(
        (savedDriverCode): boolean => savedDriverCode !== driverCode,
      ),
    });
  };

  saveFavoritesDrivers(): void {
    localStorage.setItem(
      'savedDrivers',
      JSON.stringify(this.state.savedDrivers),
    );
  }

  componentDidUpdate() {
    this.saveFavoritesDrivers();
  }

  componentDidMount() {
    this.load();
  }

  componentWillUnmount() {
    this.saveFavoritesDrivers();
    this.setState({});
  }

  private load(timeout = 50): void {
    if (this.state.isUpdating) {
      return;
    }

    this.setState({ isUpdating: true });

    clearTimeout(this.loadingTimerId);
    this.loadingTimerId = Number(
      setTimeout((): void => {
        getSeasonsResults(this.props.season, this.props.round)
          .then(data => {
            this.setState({
              racesResult: data.MRData.RaceTable.Races[0],
              isUpdating: false,
            });
          })
          .catch(error => {
            alert(error);
            this.setState({
              isUpdating: false,
            });
          });
      }, timeout),
    );
  }

  render() {
    return (
      <StandingsStyled>
        <BackButton onClick={this.props.toggleViews}>
          &larr; Back to seasons
        </BackButton>
        {this.state.isUpdating && (
          <div style={{ textAlign: 'center', marginTop: 40 }}>
            Loading results...
          </div>
        )}
        {this.state.racesResult && (
          <h2>{this.state.racesResult.Circuit.circuitName}</h2>
        )}
        {this.state.racesResult ? (
          <StandingsTable>
            <table>
              <thead>
                <tr>
                  <th>Position</th>
                  <th>Number</th>
                  <th>Driver</th>
                  <th>Code</th>
                  <th>Favorites</th>
                </tr>
              </thead>
              <tbody>
                {this.state.racesResult &&
                  this.state.racesResult.Results.map((result: Result) => {
                    const { Driver } = result;
                    const isDriverFavorite: boolean =
                      this.state.savedDrivers.includes(Driver.driverId);

                    return (
                      <StandingsTableRow
                        key={Driver.driverId}
                        position={result.position}
                        onClick={
                          isDriverFavorite
                            ? this.removeDriverFromFavorites.bind(
                                this,
                                Driver.driverId,
                              )
                            : this.addDriverToFavorite.bind(
                                this,
                                Driver.driverId,
                              )
                        }
                      >
                        <td>{result.position}</td>
                        <td>{result.number}</td>
                        <td>
                          {Driver.givenName} {Driver.familyName}
                        </td>
                        <td>{Driver.driverId}</td>
                        <td>
                          <FavoriteButton isDriverFavorite={isDriverFavorite} />
                        </td>
                      </StandingsTableRow>
                    );
                  })}
              </tbody>
            </table>
          </StandingsTable>
        ) : (
          !this.state.isUpdating && (
            <div style={{ textAlign: 'center', marginTop: 40 }}>
              No results available
            </div>
          )
        )}
      </StandingsStyled>
    );
  }
}

export default SeasonResults;
