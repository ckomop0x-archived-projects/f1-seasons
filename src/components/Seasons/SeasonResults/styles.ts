import styled from '@emotion/styled';
import Link from 'next/link';

interface StandingsTableRawTypes {
  position: any;
}

interface FavoriteButtonTypes {
  isDriverFavorite: boolean;
}

export const StandingsStyled = styled.div`
  overflow: hidden;
  width: 100%;

  h2 {
    font-family: ${({ theme }) => theme.secondaryFont};
    text-align: center;
    font-size: 1.5rem;
  }
`;

export const StandingsTable = styled.div`
  overflow-y: scroll;
  max-width: ${({ theme }) => theme.sizes.grid * 120}px;
  margin: auto;

  table {
    text-align: left;
    width: 100%;
    margin-bottom: ${({ theme }) => theme.sizes.grid * 6}px;
    border: 1px solid ${({ theme }) => theme.gray1};
    border-spacing: 0;
    border-collapse: collapse;

    th {
      font-family: ${({ theme }) => theme.secondaryFont};
      background: ${({ theme }) => theme.gray1};
      padding: ${({ theme }) => theme.sizes.grid * 2}px;
      font-size: 1rem;
      color: white;
    }

    th:nth-of-type(1),
    th:nth-of-type(2),
    th:nth-of-type(4),
    th:nth-of-type(5) {
      width: ${({ theme }) => theme.sizes.grid * 5}px;
    }
  }
`;

export const StandingsTableRow = styled.tr`
  background: ${({ position }: StandingsTableRawTypes) =>
    position === '1' ? 'rgba(255,255,0,0.4);' : ''};

  :hover {
    cursor: pointer;
    background: rgba(255, 255, 0, 0.2);
  }

  td {
    padding: ${({ theme }) => theme.sizes.grid}px;
    border-bottom: 1px solid ${({ theme }) => theme.gray1};

    :nth-of-type(1),
    :nth-of-type(2),
    :nth-of-type(4),
    :nth-of-type(5) {
      text-align: center;
    }
  }
`;

export const FavoriteButton = styled.button`
  border: none;
  width: ${({ theme }) => theme.sizes.grid * 3}px;
  height: ${({ theme }) => theme.sizes.grid * 3}px;
`;

export const BackButton = styled(Link)`
  display: block;
  margin: auto;
  border: none;
  background: none;
  cursor: pointer;
  color: ${({ theme }) => theme.colors.primaryColor};

  :hover {
    text-decoration: underline;
  }
`;
