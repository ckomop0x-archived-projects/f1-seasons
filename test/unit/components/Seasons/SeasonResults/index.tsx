import * as React from "react";
import * as renderer from 'react-test-renderer';
import SeasonResults from "../../../../../src/scripts/components/Seasons/SeasonResults";

it("renders the heading", () => {
    const props = {
        season: '2018',
        round: '1',
        year: '2018',
        toggleViews: jest.fn()
    };
    const wrapper = renderer.create(<SeasonResults {...props}/>).toJSON();

    expect(wrapper).toMatchSnapshot()
});