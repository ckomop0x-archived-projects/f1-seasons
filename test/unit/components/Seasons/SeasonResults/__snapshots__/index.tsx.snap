// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the heading 1`] = `
<div
  className="seasonStandings"
>
  <button
    className="seasonCardButton"
    onClick={[MockFunction]}
    title="Read more on Wiki"
  >
    Back to seasons
  </button>
  <table
    className="seasonStandingsTable"
  >
    <thead>
      <tr>
        <th>
          Position
        </th>
        <th>
          number
        </th>
        <th>
          Driver
        </th>
        <th>
          Code
        </th>
        <th>
          Favorites
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
